test/Test/Data/Mutable/Array.hs

```diff
@@ -1,16 +1,112 @@
+{-# LANGUAGE LinearTypes #-}
 {-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE TemplateHaskell #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
 
-module Test.Data.Mutable.Array (mutArrTests) where
+-- |
+-- Tests for mutable arrays.
+--
+-- TODO:
+--  * Test failures for out of bound access
+module Test.Data.Mutable.Array
+  ( mutArrTests,
+  )
+where
 
+import qualified Data.Array.Mutable.Linear as Arr
+import Data.Unrestricted.Linear
 import Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import qualified Prelude.Linear as Lin
+import Test.Tasty
+import Test.Tasty.Hedgehog (testProperty)
+import Test.Util
 
-mutArrTests :: IO Bool
-mutArrTests = checkParallel group
+-- # Exported Tests
+--------------------------------------------------------------------------------
 
-group :: Group
-group = Group "Mutable array tests" $
-  []
+mutArrTests :: TestTree
+mutArrTests = testGroup "Mutable array tests" group
 
+group :: [TestTree]
+group =
+  [ testProperty "[read i (write arr i x) = x]" testWrite,
+    testProperty "[alloc creates constant arrays]" testAlloc,
+    testProperty "[resize works correctly]" testResize
+  ]
 
+-- # Internal Library
+--------------------------------------------------------------------------------
 
+type ArrayTester a = Arr.Array a #-> Unrestricted Bool
```

Comment:
If you were to stick to bool, then you wouldn't need `Unrestricted`, since `Bool` is movable.

However, it's probably not the best idea, better to move to `Unrestricted (m ())` to make it possible for you to use `===`, which will give you better error message in case the test fails.

Quite frankly, I don't think this type synonym is particularly useful. You should probably just inline it.